test/antrea/networkpolicy_controller.dl

```diff
@@ -0,0 +1,364 @@
+    pod in k8s.Pod(.namespace = Some{ns}),
+    pod.status.podIP != "",
+    k8s.labelSelectorMatches(addressGroup.selector.podSelector, pod.labels).
+
+AddressGroupAddress(addressGroup.uid, pod.status.podIP) :-
+    // Pods must be selected from Namespaces matching nsSelector.
+    addressGroup in AddressGroup(.selector = GroupSelector{.nsSelector = NSSelectorLS{namespaceSelector}, .podSelector = podSelector}),
+    namespace in k8s.Namespace(),
+    k8s.labelSelectorMatches(Some{namespaceSelector}, namespace.labels),
+    pod in k8s.Pod(.namespace = Some{namespace.name}),
+    pod.status.podIP != "",
+    is_none(podSelector) or k8s.labelSelectorMatches(podSelector, pod.labels).
+
+// NetworkPolicyAddressGroup: stores all address groups that a network policy references;
+// used to compute AddressGroupSpan.
+relation NetworkPolicyAddressGroup(np: k8s.UID, addressGroup: string)
+
+NetworkPolicyAddressGroup(np, addressGroup) :-
+    NetworkPolicy(.uid = np, .rules = rules),
+    var addressGroups = {
+        var addressGroups: Vec<string> = vec_empty();
+        for (rule in rules) {
+            vec_append(addressGroups, rule.from.addressGroups);
+            vec_append(addressGroups, rule.to.addressGroups)
+        };
+        addressGroups
+    },
+    var addressGroup = FlatMap(addressGroups).
+
+AddressGroupSpan(k8s.UID{addressGroup}, span) :-
+    AddressGroup(.uid = k8s.UID{addressGroup}),
+	// Get all internal NetworkPolicy objects that refers this AddressGroup.
+    NetworkPolicyAddressGroup(np, addressGroup),
+    NetworkPolicySpan(np, npSpan),
+    var span = Aggregate((addressGroup), group_set_unions(npSpan)).
+
+// AppliedToGroupPod: pods that belong to a group along with nodes
+// where the pod is scheduled;
+// used to compute AppliedToGroupPodsByNode and AppliedToGroupSpan.
+relation AppliedToGroupPod(appliedToGroup: k8s.UID, pod: k8s.UID, nodeName: string)
+
+AppliedToGroupPod(appliedToGroup, pod.uid, pod.spec.nodeName) :-
+    // XXX: Is it correct that we only consider selectors with namespace name set?
+    AppliedToGroup(.uid = appliedToGroup,
+                   .selector = GroupSelector{.nsSelector = NSSelectorNS{namespace}, .podSelector = podSelector}),
+	// Retrieve all Pods matching the podSelector.
+    pod in k8s.Pod(.namespace = Some{namespace}),
+    // No need to process Pod when it's not scheduled.
+    pod.spec.nodeName != "",
+    k8s.labelSelectorMatches(podSelector, pod.labels).
+
+AppliedToGroupPodsByNode(appliedToGroup, podsByNode) :-
+    AppliedToGroupPod(appliedToGroup, pod, nodeName),
+    var podsOnNode = Aggregate((appliedToGroup, nodeName), group2set(pod)),
+    var podsByNode = Aggregate((appliedToGroup), group2map((nodeName, podsOnNode))).
+
+AppliedToGroupSpan(appliedToGroup, span) :-
+    AppliedToGroupPod(appliedToGroup, pod, _),
+    var span = Aggregate((appliedToGroup), group2set(pod)).
+
+NetworkPolicySpan(policy, span) :-
```

Comment:
it's remarkable how compact these rules are. they are actually easier to read than the comments that describe them...